nft-qos: add new package

rosysong opened the pull request:
This is the nftables implementation for qos limit rate

Signed-off-by: Rosy Song <rosysong@rosinson.com>

![nqos-examples](https://user-images.githubusercontent.com/37688994/40963659-c1a72cd2-68db-11e8-87e7-f82a7ae8544f.png)

luci-app is pulled at https://github.com/openwrt/luci/pull/1855

hnyman commented on net/nqos/files/nqos.config:
```diff
@@ -0,0 +1,17 @@
+#
```
Default config should have at least some default rules uncommented. Right now the example may be a bit confusing, as it is not obvious which options are mandatory. 

And the example does not say what is the default for optional "unit")

It would also be great to have an "enabled" switch 0/1 in the config file, and defaulting to 0, so that the package can be installed in the image and still be disabled by default.

rosysong commented:
@hnyman  Hi, thank you for your suggestions, I updated my pull request, please check it again.

rosysong commented:
By the way, if we want to have a global speed limit, just set the ip for the network with netmask, e.g. 192.168.1.0/24,  172.168.6.0/24,

rosysong commented:
Hi @hnyman  This is my new commit, This a new nft-qos, 

1. Add static qos/ auto qos support
2. Add traffic shaping support

Static QoS
![static-qos](https://user-images.githubusercontent.com/37688994/42809432-3b2ac258-89e8-11e8-9c40-8cac728aa027.png)


Auto Qos
![dynamic-qos](https://user-images.githubusercontent.com/37688994/42809458-4a469e4c-89e8-11e8-868b-6cda0865e74d.png)


 Traffic Shaping
![traffic-control](https://user-images.githubusercontent.com/37688994/42809477-52662a48-89e8-11e8-8aa0-208d83c2470e.png)

Please don't hesitate to give me more suggestions. thanks!!

rosysong commented:
This package needs kmod-nft-netdev which was pulled at https://github.com/openwrt/openwrt/pull/1189

rosysong commented:
Hi @hnyman , the dependencies of the package were merged (openwrt/openwrt#1189, openwrt/openwrt#1231), what's left me to do ? :-)

hnyman commented:
would be great if some firewall / iptables / SQM experts like @jow- or @moeller0 or @ldir-EDB0   would glance at your code.

moeller0 commented:
So, I see this as complimentary to sqm-scripts. The one thing I will do once I find time is to test how this affects latency and throughput. Personally I am just fine with cake's per-IP fairness, but there are repeated requests for a finer grained bandwidth management for specific hosts in the context of the qos/sqm threads in the forum, and I believe that this is an somewhat orthogonal problem to wan bufferbloat mitigation, so I am quite happy to see this as a dedicated separate package.

rosysong commented:
@hnyman @jow- @moeller0 @ldir-EDB0 Hi, this is the new commit for nft-qos,
ChangeLog:

1. rework for API scripts and bugs fixed, available whitelist in static-qos either.

Please feel free to give me more suggestions, thanks!

rosysong commented:
@hnyman @jow- @moeller0 @ldir-EDB0 Hi, here is the new commit for nft-qos,
ChangeLog:

1. fix typo error
2. fix dependency issue on IPv6

Thanks!

hnyman commented:
There has been not that many comments, but as Travis is happy, I guess that this could be merged.

However, the buildbot will never build this into the download repo, as there is config dependency for a non-default nft config option PACKAGE_NFT_WITH_JSON . Buildbot does not have that option set, so this package will never get built in buildbot. 

So, the usage of this package would be quite minimal, as somebody needs to find this and compile from source along a modified nftables.

moeller0 commented:
> would be great if some firewall / iptables / SQM experts like @jow- or @moeller0 or @ldir-EDB0 would glance at your code.

I know I am coming a bit late to the party here, but better late than never, I guess... Now, from my current perspective I see no interactions with sqm here, neither negative (nor positive). I do notice that the functionality to throttle individual machines/ip-addresses often comes up in sqm related discussions so the this capability should make a lot of user's happy, especially if it can be seamlessly combined with sqm. Now, I should actually go and measure if this approach is sufficiently bufferbloat free to be recommended unconditionally to all gamer-types.
Now, I have a silly question, what do I need to do test this? Do I need to build from source, or can I just install required packages from a repository?

rosysong commented:
@hnyman  @moeller0  I updated my commit, fixed the dependency issue but needs below patch :

diff --git a/package/network/utils/nftables/Makefile b/package/network/utils/nftables/Makefile
index d1e995116d..1d9cae7c4d 100644
--- a/package/network/utils/nftables/Makefile
+++ b/package/network/utils/nftables/Makefile
@@ -42,6 +42,7 @@ define Package/nftables/config
        config PACKAGE_NFT_WITH_JSON
                bool "Build nftables with json support"
                depends on PACKAGE_nftables
**+               default y if PACKAGE_nft-qos**
                default n
 endef
 
After apply this patch, once you select nft-qos, nftables and PACKAGE_NFT_WITH_JSON will be choosed as well.

hnyman commented:
Sorry, but I think that this would not work in reality. Sounds like a circular dependency (and between sources repos).

Explanation: PACKAGE_NFT_WITH_JSON get automatically selected if PACKAGE_nft-qos gets selected. But  PACKAGE_nft-qos is not even visible unless PACKAGE_NFT_WITH_JSON  is already selected.

Just try with empty .config and then go into menuconfig.

Regarding buildbot it gets even more complicated, as core packages (including nftables) are built in phase1 buildbot that does not know about extra packages. So PACKAGE_nft-qos is not set there, I think, (as the whole nft-qos is in another feed repo). Add-on packages like nft-qos are then built in phase2 buildbot, but it does not touch the kernel-related packages build in phase1.

So, the dependency on a non-default package option is really troublesome.

rosysong commented:
@hnyman  Thanks for your explanation, that's why I show the patch at here but not in my current PR.
The troublesome dependency does not added in nft-qos.

champtar commented:
@rosysong What is the old / new size of nftables package when enabling PACKAGE_NFT_WITH_JSON ?
what is the size of the additionnal dependencies ?
I would argue that almost nobody uses nftables, so changing PACKAGE_NFT_WITH_JSON default to yes can be ok (people that do use nftables are likely to compile OpenWRT themselves)

rosysong commented:
@hnyman  Thanks, here is the previous commit.

@champtar  I think making make nftables to have two different variants is a good idea as people can choose which variant they want. I have just opened a new pull request(https://github.com/openwrt/openwrt/pull/1454) for this. Once the new PR of nftables is merged, I will update the nft-qos dependency as well.

rosysong commented:
Any plan to merge it? Thanks

champtar commented on net/nft-qos/Makefile:
```diff
@@ -0,0 +1,58 @@
+define Package/nft-qos
+  SECTION:=utils
+  CATEGORY:=Base system
+  DEPENDS:=@PACKAGE_NFT_WITH_JSON +kmod-nft-netdev +kmod-nft-bridge
```
missing `+nftables`

champtar commented:
@rosysong I know you are waiting for a long time but the idea is to have packages that work out of the box, so we should wait for https://github.com/openwrt/openwrt/pull/1454 to be solved / merged

rosysong commented:
@champtar Only 8 bytes increased in comparation with the ipk without json(with jansson) support.

hnyman commented:
> Only 8 bytes increased in comparation with the ipk without json(with jansson) support.

Plus the implicit increase of always installing also jansson itself.

rosysong commented:
@hnyman  The size of jansson is 19k

```
ls -l bin/packages/mips_24kc/packages/jansson_2.11-1_mips_24kc.ipk -h
-rw-r--r-- 1 xxx xxx 19K Nov  2 09:10 bin/packages/mips_24kc/packages/jansson_2.11-1_mips_24kc.ipk
```

rosysong commented:
Hi, all 
   Here is the newer commit,
   I would like to tell you a good news, the nft-qos now can work without json support on nftables, let's enjoy it. Thanks!!!  :-)   Please check it again.   :-)
